[WIP] The Big Freeze

duncathan opened the pull request:
This is a work in progress and nowhere near complete right now. I'm just making the PR for sake of convenience and to let everyone see it.
Concept from https://tgstation13.org/phpBB/viewtopic.php?f=9&t=4623
Yes, I intend to be accepting payment from J_Madison when I have completed this work. Just disclosing that here.

As it is I've really only just started; the meat of this isn't even begun.
Proposed scaling: https://www.desmos.com/calculator/kh3e305nzf
TODO:
- [x] Set up the game_mode foundation
- [ ] Set up a working objective (cannot yet be done)
- [ ] Pawn abilities (should they have any)
- [ ] Scion abilities (names not final)
  - [x] Pre-transform
    - [x] Spread Frost (similar to alien weeds, except it makes shit colder)
    - [x] Chilling Grasp (think enthrall except re-fluffed)
  - [ ] Transform (similar to shadowling hatch)
  - [x] Post-transform
    - [x] Freeze Area (rapidly cools area around them)
    - [x] Extinguish (puts out fires and removes plasma (replace it with cold n2?))
    - [x] Frostbite (purges cold-res effects from a target)
    - [x] Re-freeze (heals pawns and perhaps self)
- [ ] Ability logging and player feedback
- [ ] Equipment
  - [ ] Armor
    - [ ] Regalia (scion armor)
    - [ ] Pawn armor
  - [x] Shields
  - [ ] Weapons (lower target's body temp; melts when exposed to heat or salt)
    - [x] Lance (pawn weapon)
    - [ ] Sceptre (scion weapon)
- [x] Scion code
  - [x] Heat and burn sensitivity
  - [x] Heal ~~around frost~~ when exposed to cold environments
  - [ ] Passive energy weapon drain (if feasible)
- [x] Pawn code
  - [ ] Passive energy weapon drain (if feasible)
- [ ] Antag HUD
- [x] Sprites (MOSTLY DONE, THANKS TO NIENHAUS)
- [ ] Whatever else J_Madison asks for

Sprites are absolutely needed; ask J_Madison for details on what's needed in that department. If I must make them I will, but it sure won't be pretty.

REMINDER: while it doesn't mean a ton to me whether this PR is merged, any feedback received will also be received by J_Madison, the client. I'm certain that they want this gamemode to be good and fun, and I will do what I can to ensure that it will be both of those things. Ultimately, I would like to see this merged in some form or another.

duncathan committed:
preliminary work on big freeze

Xhuis commented on code/game/gamemodes/freeze/freeze.dm:
```diff
@@ -0,0 +1,79 @@
+		var/explanation
+		switch(frost_objectives[i])
+			if("freeze")
+				explanation = "man I don't fuckin know talk to J_Madison"
```
top kek

Incoming5643 commented on code/game/gamemodes/freeze/freeze.dm:
```diff
@@ -0,0 +1,79 @@
+		log_game("[scion.key] (ckey) has been selected as a Scion of the Kingdom.")
+		sleep(10)
+		scion.current << "<br>"
+		scion.current << "<span class='shadowling'><b><font size=3>You are a shadowling!</font></b></span>"
```
:cop: :oncoming_police_car: :cop: :oncoming_police_car: :cop:
:copyright: :moneybag:  :copyright: :moneybag: :copyright:
:cop: :oncoming_police_car: :cop: :oncoming_police_car: :cop:

duncathan replied:
sh-shush

duncathan committed:
chilling grasp and create frost abilities

Anonus commented on code/game/gamemodes/freeze/freeze.dm:
```diff
@@ -0,0 +1,89 @@
+	spawn(0)
+		if(scion_mind.assigned_role == "Clown")
+			var/mob/living/carbon/human/S = scion.current
+			S << "<span class='notice'>Your REALLY COLD, LIKE, YOU'RE LITERALLY ICE nature has allowed you to overcome your clownishness.</span>"
```
ICE COLD

duncathan replied:
ABSOLUTE ZERO

Miauw62 replied:
https://www.youtube.com/watch?v=FSZ25ikUKVY

Cheridan commented:
if i make sprites do i get a cut of the dosh @duncathan

RemieRichards commented on code/game/gamemodes/freeze/structures.dm:
```diff
@@ -0,0 +1,43 @@
+//turf frost - these are children of alien weeds because they're basically identical
+//exposure to air above T20C will slowly melt them, wetting the floor as they melt
+/obj/structure/alien/weeds/frost
```
you know it's kind of freaky how similar my implementation would have been to yours.

Miauw62 replied:
i laughed when i read this

RemieRichards commented on code/game/objects/effects/aliens.dm:
```diff
@@ -183,8 +188,9 @@
-	for (var/obj/structure/alien/weeds/W in range(1,T))
-		W.updateWeedOverlays()
+	for(var/obj/sctructure/alien/weeds/W in range(1,T))
+		if(istype(W, type_of_weed)) //so that you don't get weeds connecting to frost
```
remove the specific nature of the comment.
just mention you don't get mixing types.

duncathan commented:
@Cheridan Payment is determined by J_Madison; talk to them.

ike709 commented:
sounds cool but when i first saw the name of this pr i thought it was a feature freeze

duncathan committed:
frost part 2; frost is now finished excluding spritework

duncathan committed:
species code part 1

duncathan commented:
Wow I am impressed with how easy it was to do the whole species thing

LanCartwright commented:
Why does this remind me of Yuki-onna?

RemieRichards commented:
@LanCartwright 
Yuki means snow, this is an ice/snow themed update
Onna means woman, we do have female players and characters

Reasonable link to make.

duncathan committed:
proc for checking if one can use a given scion ability

duncathan committed:
more abilities: freeze area and frostbite. they are lacking player feedback.
also adds ""damage type"" COLD (applies burn damage but is affected by coldmod)

duncathan commented:
I've added a couple more abilities, these ones for a transformed scion (haven't made the transform ability yet)
I've also added the new ""damage type"" (it's not actually a damage type per se but it can be passed to apply_damage and can be used as an object's damage type (I think)) COLD that will apply burn damage based multiplied by the species' coldmod. This causes a highly situational balance change in which species with a higher burnmod than coldmod (shadowlings notably) will take slightly less damage in a cold environment, while species with a higher coldmod than burnmod (do they exist?) will take more.

duncathan committed:
abilities: extinguish and re-freeze

duncathan commented:
All right, since I've finished up all the abilities except for transform, I figured I should go over what each one does in detail.
#Scion Abilities
##Pre-transform
These can be used before going loud.
###Spread Frost
Plants frost underneath caster
10 second cooldown

So what we've got here is, essentially, alien weeds. Except they cool the area around them, and make the floor slippery when they melt. They melt when exposed to too much heat (I think they begin to melt at temperatures above T20C) and when attacked with welders.
They cool their turf to (T20C - temperature_delta) degrees. temperature_delta is currently 15, making their target temperature 5 degrees C.
They are not currently slippery in their normal state.
###Chilling grasp
Target adjacent to caster
No cooldown, but takes time to cast.

I'm gonna be completely honest, this is Enthrall re-fluffed. It also checks for frost kingdom stuff, not shadowling stuff, and will only work if the victim is not wearing a jumpsuit.
##Transform
Not yet completed.
##Post-transform
For when you've gone loud.
###Freeze Area
3x3 area centered on caster
30 second cooldown

This one cools the area around the scion by 30 degrees.
###Extinguish
Area centered on caster
60 second cooldown

This plants a fluffed up nanofrost at the caster's feet. Being nanofrost, it replaces combusting plasma in the air with cold N2 and chills the area.
###Frostbite
Target adjacent to caster
30 second cooldown

This purges cold-resistant effects in the target, and deals cold damage inversely proportional to how many types of effects were purged. If the target had the cold resistance mutation and had inaprovaline in their body, no damage will be dealt. If the target had only one of those, 10 damage will be dealt and the target's body temperature will be lowered by 50 degrees (subject to change). If neither, 20 damage will be dealt and body temperature will be lowered by 100 degrees (wow).
###Re-freeze
Target adjacent to caster
10 second cooldown

If the target is an ally (pawn or scion) this will deal 30 COLD damage (thus healing pawns by 15 and healing scions by 30) and will lower the target's body temperature by 40 degrees.

Anonus commented:
I think most of the single-target one-tile-range things should use, like, a frost hand object a la ei nath/flesh to stone, because FUCK the spell interface.

Xhuis commented:
can you play freeze tag

Iamgoofball commented on code/modules/projectiles/projectile/special.dm:
```diff
@@ -46,7 +46,7 @@
 	icon_state = "ice_2"
 	damage = 0
-	damage_type = BURN
+	damage_type = COLD
```
You need to add SetColdDamage() and AdjustColdDamage(), and make health and all other factors count in Cold damage.

duncathan replied:
>I've also added the new ""damage type"" (it's not actually a damage type per se but it can be passed to apply_damage and can be used as an object's damage type (I think)) COLD that will apply burn damage based multiplied by the species' coldmod. This causes a highly situational balance change in which species with a higher burnmod than coldmod (shadowlings notably) will take slightly less damage in a cold environment, while species with a higher coldmod than burnmod (do they exist?) will take more.

It applies burn damage, it's simply affected by coldmod. COLD can be passed as an argument to apply_damage().

Iamgoofball replied:
rip, thought we had an actual cold damage type now

duncathan committed:
changes to abilities

duncathan committed:
prelimary work for frost equipment; adds nienhaus' sprites

duncathan committed:
makes the game compile

duncathan commented on code/game/gamemodes/freeze/scion_abilities.dm:
```diff
@@ -0,0 +1,261 @@
+		target.reagents.del_reagent("inaprovaline")
+		affected++
+	var/damageToAfflict = base_dmg - ((affected*0.5)*base_dmg) //does no damage if we removed two effects; half damage if we removed one effect; max damage if we removed none
+	target.apply_damage(damageToAfflict, COLD)
```
so this works fine...

duncathan commented on code/game/gamemodes/freeze/scion_abilities.dm:
```diff
@@ -0,0 +1,261 @@
+	if(!is_frosty(target))
+		return 0
+
+	target.apply_damage(base_dmg/*, COLD*/)
```
...but this throws a "bad argument definition" compile error when I uncomment COLD.
what the hell am I doing wrong?

Xhuis replied:
What's the error?

RemieRichards replied:
@Xhuis "bad argument definition", as he said.

Xhuis replied:
Oh. Sorry, just woke up.
@duncathan You might want to put up a PiratePad and get some people on Coderbus to help you out. I'm about to leave, but if it still isn't resolved by then I'll take a closer look.

duncathan committed:
i did a lot of changes this commit and tbh i'm not sure what exactly it consists of

duncathan commented:
For some reason humans go suddenly bald when made a scion. Not sure what's causing that, but I had a good chuckle at the thought of fucking hair being used for metagaming

Xhuis commented:
Do scions have a mutant race? That's why, I had the same problem with early iterations of thralls.

duncathan committed:
gives frosties COLD_RES and changes how they heal; ensures that aliens don't heal on frost; adds proc set_mrace_keep_values
set_mrace_keep_values(mob/living/carbon/owner, datum/species/mrace)
  calls hardset_dna with all proper values to maintain all info about the owner (ui, ue, se, etc)

duncathan commented:
I'm still making progress on this. Most of the work has involved trying to fix the baldness thing due to mutantraces (whyyyy) and making all the equipment.
Once that's sorted out the last main things will be the objective (yet undecided) and the antag HUD.

duncathan committed:
tiny tiny update

WJohn commented:
What's the status on this?

duncathan commented:
Work is happening, just far more slowly since school has started back up.

duncathan committed:
I can't believe that this whole time there was a species flag for hair

duncathan commented:
who knew that there were appearance flags for mutantraces

Razharas commented:
@duncathan

JJRcop commented:
@duncathan

duncathan commented:
wtf I swear it's been less than two weeks since I touched this last

MrStonedOne commented:
Please make a changelog,

You can add it to the PR's body rather then as a file now, see here: https://tgstation13.org/phpBB/viewtopic.php?f=5&t=5157

duncathan commented:
>16 days ago

God damn what happened

Steelpoint commented:
Time flies.

Aranclanos commented:
inactivity

Iamgoofball commented:
@duncathan get 2 work
